src/ring_theory/tensor_product.lean

```diff
@@ -1,43 +1,245 @@
+- Define the `R`-algebra structure on `A ⊗ B`, for another `R`-algebra `B`.
+  and provide the structure isomorphisms
+  * `R ⊗[R] A ≃ₐ[R] A`
+  * `A ⊗[R] R ≃ₐ[R] A`
+  * `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+  * `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
+
+## Main declaration
+
+- `linear_map.base_change A f` is the `A`-linear map `A ⊗ f`, for an `R`-linear map `f`.
 
-The code for
-* `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
-is written and compiles, but takes longer than the `-T100000` time limit,
-so is currently commented out.
 -/
 
 universes u v₁ v₂ v₃ v₄
 
-namespace algebra
-
 open_locale tensor_product
 open tensor_product
 
 namespace tensor_product
 
+variables {R A M N P : Type*}
+
+/-!
+### The `A`-module structure on `A ⊗[R] M`
+-/
+
+open linear_map
+open algebra (lsmul)
+
+namespace algebra_tensor_module
+
+section semiring
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [module R M] [module A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [module R N]
+variables [add_comm_monoid P] [module R P] [module A P] [is_scalar_tower R A P]
+
+lemma smul_eq_lsmul_rtensor (a : A) (x : M ⊗[R] N) : a • x = (lsmul R M a).rtensor N x := rfl
+
+/-- Heterobasic version of `tensor_product.curry`:
+
+Given a linear map `M ⊗[R] N →[A] P`, compose it with the canonical
+bilinear map `M →[A] N →[R] M ⊗[R] N` to form a bilinear map `M →[A] N →[R] P`. -/
+@[simps] def curry' (f : (M ⊗[R] N) →ₗ[A] P) : M →ₗ[A] (N →ₗ[R] P) :=
+{ map_smul' := λ c x, linear_map.ext $ λ y, f.map_smul c (x ⊗ₜ y),
+  .. curry (f.restrict_scalars R) }
+
+lemma restrict_scalars_curry' (f : (M ⊗[R] N) →ₗ[A] P) :
+  restrict_scalars R (curry' f) = curry (f.restrict_scalars R) :=
+rfl
+
+/-- Just as `tensor_product.mk_compr₂_inj` is marked `ext` instead of `tensor_product.ext`, this is
+a better `ext` lemma than `tensor_product.algebra_tensor_module.ext` below.
+
+See note [partially-applied ext lemmas]. -/
+@[ext] lemma curry'_injective :
+  function.injective (curry' : (M ⊗ N →ₗ[A] P) → (M →ₗ[A] N →ₗ[R] P)) :=
+λ x y h, linear_map.restrict_scalars_injective R $ curry_injective $
+  (congr_arg (linear_map.restrict_scalars R) h : _)
+
+theorem ext {g h : (M ⊗[R] N) →ₗ[A] P}
+  (H : ∀ x y, g (x ⊗ₜ y) = h (x ⊗ₜ y)) : g = h :=
+curry'_injective $ linear_map.ext₂ H
+
+end semiring
+
+section comm_semiring
+variables [comm_semiring R] [comm_semiring A] [algebra R A]
+variables [add_comm_monoid M] [module R M] [module A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [module R N]
+variables [add_comm_monoid P] [module R P] [module A P] [is_scalar_tower R A P]
+
+/-- Heterobasic version of `tensor_product.lift`:
+
+Constructing a linear map `M ⊗[R] N →[A] P` given a bilinear map `M →[A] N →[R] P` with the
+property that its composition with the canonical bilinear map `M →[A] N →[R] M ⊗[R] N` is
+the given bilinear map `M →[A] N →[R] P`. -/
+@[simps] def lift' (f : M →ₗ[A] (N →ₗ[R] P)) : (M ⊗[R] N) →ₗ[A] P :=
```

Comment:
Given you're not in the `tensor_product` namespace here, should this just be called `lift` without the `'`?